source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs

```diff
@@ -140,11 +140,43 @@ private void SetOutputVariables(RunningDeployment deployment, IEnumerable<S3Uplo
+                log.Info($"Saving object ARN to variable \"Octopus.Action[{actionName}].Output.Package.Arn\"");
+                log.SetOutputVariableButDoNotAddToVariables($"Package.Arn", $"arn:{(awsEnvironmentGeneration.AwsRegion.SystemName.Equals("cn-north-1") ? "aws-cn" : "aws")}:s3:::{successfulResults[0].BucketName}/{successfulResults[0].BucketKey}");
+                log.Info($"Saving object version id to variable \"Octopus.Action[{actionName}].Output.Package.ObjectVersion\"");
+                log.SetOutputVariableButDoNotAddToVariables($"Package.ObjectVersion", successfulResults[0].Version);
+                log.Info($"Saving object version id to variable \"Octopus.Action[{actionName}].Output.Files[{successfulResults[0].BucketKey}]\"");
+                log.SetOutputVariableButDoNotAddToVariables($"Files[{successfulResults[0].BucketKey}]", successfulResults[0].Version);
+            }
+            else
+            {
+                foreach (var result in successfulResults)
+                {
+                    var fileName = Path.GetFileName(result.BucketKey);
+                    log.Info($"Saving bucket key to variable \"Octopus.Action[{actionName}].Output.Package.Key[{fileName}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Package.Key[{fileName}]", result.BucketKey);
+                    log.Info($"Saving object S3 URI to variable \"Octopus.Action[{actionName}].Output.Package.S3Uri[{fileName}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Package.S3Uri[{fileName}]", $"s3://{result.BucketName}/{result.BucketKey}");
+                    log.Info($"Saving object URI to variable \"Octopus.Action[{actionName}].Output.Package.Uri[{fileName}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Package.Uri[{fileName}]", $"https://{result.BucketName}.s3.{awsEnvironmentGeneration.AwsRegion.SystemName}.amazonaws.com/{result.BucketKey}");
+                    //  ARN format: `arn:aws:s3:::bucket-name/key` (Note: China (Beijing region (cn-north-1) uses `aws-cn` instead of `aws`)
+                    log.Info($"Saving object ARN to variable \"Octopus.Action[{actionName}].Output.Package.Arn[{fileName}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Package.Arn[{fileName}]", $"arn:{(awsEnvironmentGeneration.AwsRegion.SystemName.Equals("cn-north-1") ? "aws-cn" : "aws")}:s3:::{result.BucketName}/{result.BucketKey}");
+                    log.Info($"Saving object version id to variable \"Octopus.Action[{actionName}].Output.Package.ObjectVersion[{fileName}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Package.ObjectVersion[{fileName}]", result.Version);
+                    log.Info($"Saving object version id to variable \"Octopus.Action[{actionName}].Output.Files[{result.BucketKey}]\"");
+                    log.SetOutputVariableButDoNotAddToVariables($"Files[{result.BucketKey}]", result.Version);
+                }
```

Comment:
Could you join these two blocks? I think they differ only by the filename key you add at the end, so you could always append that postfix, but for the case where there is only one element leave the postfix as an empty string.